### ISSUE TITLE
Slow convergence when overfitting CUT3R Stage 1 on a small set with fixed encoder

### ISSUE THREAD

**Issue by ChaoyiZh** [CUT3R/CUT3R#75]
---

### Description

Hi, thank you for the great work! I’m trying to overfit **CUT3R Stage 1** on a small subset of TartanAir for debugging and sanity check purposes.

- **Dataset**: `hospital/easy/P001` trajectory from TartanAir (~400 frames)
- **Training goal**: Just overfit this single sequence
- **Hardware**: 4 × A100 80GB
- **Encoder**: I loaded the pretrained CUT3R encoder and froze it
- **Config**: see below
- **Training progress**: ~80 epochs trained
- **Observations**:
  - The loss does decrease, but the convergence feels slow
  - Attached below are my current loss curve and visualizations

---

### Questions

1. Is the config below appropriate for Stage 1 training in this overfitting setting?
2. If yes, is the current convergence behavior expected? Any typical number of epochs or loss threshold I should look for in this toy setup?

---

### Config Used

```yaml
model: "ARCroco3DStereo(ARCroco3DStereoConfig(freeze='encoder', state_size=768, state_pe='2d', pos_embed='RoPE100', rgb_head=True, pose_head=True,  img_size=(224, 224), head_type='linear', output_mode='pts3d+pose', depth_mode=('exp', -inf, inf), conf_mode=('exp', 1, inf), pose_mode=('exp', -inf, inf), enc_embed_dim=1024, enc_depth=24, enc_num_heads=16, dec_embed_dim=768, dec_depth=12, dec_num_heads=12))"
# pretrained: null
pretrained: cut3r_512_dpt_4_64.pth
load_only_encoder: True
long_context: False
fixed_length: True
resume: null
benchmark: True
num_views : 4
num_test_views : 4
n_corres_train: 0
n_corres_test: 0

train_criterion: ConfLoss(Regr3DPose(L21, norm_mode='?avg_dis'), alpha=0.2) + RGBLoss(MSE)
test_criterion: Regr3DPose(L21, norm_mode='?avg_dis', gt_scale=True, sky_loss_value=0) + Regr3DPose_ScaleInv(L21, norm_mode='?avg_dis', gt_scale=True, sky_loss_value=0) + RGBLoss(L21)

train_dataset: 1000 @ TartanAir_Multi(split=None, ROOT="/data/chaoyi/cut3r_data/processed_tartanair", resolution=224, num_views=${num_views}, n_corres=${n_corres_train})

test_dataset: 1000 @ TartanAir_Multi(split=None, ROOT="/data/chaoyi/cut3r_data/processed_tartanair", resolution=224, num_views=4, seed=42, n_corres=0)

seed: 0
batch_size: 16
accum_iter: 1
gradient_checkpointing: False
epochs: 100
start_epoch: 0
weight_decay: 0.05
lr: 1e-4
min_lr: 1e-6
warmup_epochs: 10
amp: 1

num_workers: 8
world_size: 1
local-rank: -1
dist_url: 'env://'
rank: 0
gpu: 0
distributed: False
dist_backend: 'nccl'

eval_freq: 1
save_freq: 1
keep_freq: 10
print_freq: 10
print_img_freq: 500
num_imgs_vis: 4
save_dir: 'checkpoints'
exp_name: 'cut3r_train_first_stage_fix_encoder'
task: 'cut3r'
logdir: ./${save_dir}/${exp_name}/logs
output_dir: ./${save_dir}/${exp_name}/
hydra:
  verbose: True
  run:
    dir: ./${save_dir}/${exp_name}
```
---
### Config Training Logs / Loss Curve

<img width="592" height="347" alt="Image" src="https://github.com/user-attachments/assets/64a0b916-1ca9-4240-94fc-20ec5ca9a7b5" />

---

### Visualizations

<img width="1162" height="614" alt="Image" src="https://github.com/user-attachments/assets/ba4b7142-4b02-4971-81d0-18baee2b8661" />

Let me know if more logs or debugging info would help. Thanks for your work on this repo!

**Comment by Forrest-110** [comment]
Hi @ChaoyiZh! Thank you for your inquiry.

Before diving into the issue, I’d like to clarify a couple of points:

1. The training loss curve you provided appears to contain only 1,284 steps, which seems unusually slow to take approximately 15 hours on A100s. Could you confirm if this is indeed the case?

2. The visualizations you shared are from step 95,000 and seem to be from validation, which would correspond to around epoch 95. Could you confirm whether this visualization is indeed from epoch 95?

**Comment by ChaoyiZh** [comment]
Hi, thank you for your response and great questions!

1. The training process was accidentally interrupted once, and after manually resuming, TensorBoard seems to have recorded incorrect timestamps. In reality, training 100 epochs typically takes around 1–2 hours on A100s.

2. Yes, the visualizations I shared are indeed from around epoch 95, which corresponds to step 95,000 in the logs.

Please let me know if further clarification is needed!

**Comment by Forrest-110** [comment]
Thank you. Also, it seems that your test set differs from your training set. Since you're trying to overfit, I would expect both to be the same—especially given that your training set only contains 1,000 samples. This mismatch might explain why your validation results appear so poor. 

If that's not the case, I would recommend double-checking your training process. The config looks good overall, aside from the dataset issue mentioned above. However, based on the visualization, the model's performance appears comparable to what one would expect after only around 10 epochs of training.

**Comment by ChaoyiZh** [comment]
Thank you for the follow-up!

I believe the train and test sets should actually be the same in my current setup. In my dataset class, I’ve hardcoded the specific trajectory to use, so based on the current config, there shouldn't be any difference between the training and test data. Please correct me if I’m mistaken.

That said, I agree with your observation—after around 10 epochs, the visualizations don’t seem to improve significantly.
